### ISSUE TITLE
Solver timeout

### ISSUE THREAD

**Issue by csH7KmCC9** [pysmt/pysmt#689]
Hi,

Thanks for this library. Works a treat!

What is the best way to set a solver timeout? I can probably provide an option to e.g. Z3 directly, but it's not clear to me exactly how to use `Solver()` with solver options. Can you please provide an example?

Thanks again!

**Comment by mikand** [comment]
Hi @csH7KmCC9, thanks for using pysmt.

We do not directly support timeouts, because not all the solvers support it and some (e.g. MathSAT) have relaxed guarantees when it comes to timeouts.

One easy thing you can try is to put the solving code in a separate process and impose a timeout. this is super-easy if you use [pebble](https://pebble.readthedocs.io/en/latest/) (Same thing is doable with a bit more effort using the `multiprocessing` standard library if you cannot install 3rd party packages).

```python
from pebble import concurrent
from concurrent.futures import TimeoutError
from pysmt.shortcuts import Solver

@concurrent.process(timeout=1) # You can adjust the timeout here
def solve(formula):
    with Solver(name='z3') as s:
        s.add_assertion(formula)
        return s.solve()

f = ... # Your formula here

future = solve(f)

try:
    print(future.result())
except TimeoutError:
    print('TIMEOUT!')
``` 

HTH

**Comment by mikand** [comment]
Ah, I forgot to mention how to use the Z3 embedded timeout via `solver_options`:

```python
from pysmt.shortcuts import *
from pysmt.exceptions import SolverReturnedUnknownResultError

x, a, b, c = (Symbol(n, REAL) for n in 'xabc')
f = ForAll([x], Equals(Plus(Times(a, Times(x, x)), Times(b, x), c), Real(0)))

for t in range(1, 50, 2):
    try:
        with Solver(name='z3', solver_options={'timeout': t}) as s:
            s.add_assertion(f)
            res = s.solve()
            print(res)
    except SolverReturnedUnknownResultError:
        print('TIMEOUT!')
```

**Comment by csH7KmCC9** [comment]
Yeah, I used multiprocessing [for a quick hack](https://github.com/csH7KmCC9/coinjoin-smt/blob/829b593259d30b4c7befce4c1267c227126bd29a/prototype.py#L220), but I figured it would be better to do it the 'proper' way. Your second comment is exactly what I was looking for. I think it would be helpful to add this to the examples directory, too.

Thanks a lot!
